scalding-core/src/main/scala/com/twitter/scalding/Operations.scala

```diff
@@ -169,24 +169,108 @@ package com.twitter.scalding {
       // Docs say it is safe to do this cast:
       // http://docs.cascading.org/cascading/2.1/javadoc/cascading/operation/Operation.html#flush(cascading.flow.FlowProcess, cascading.operation.OperationCall)
-      val functionCall = operationCall.asInstanceOf[FunctionCall[MapsideCache[V]]]
+      val functionCall = operationCall.asInstanceOf[FunctionCall[MapsideCache[Tuple, V]]]
       val cache = functionCall.getContext
       add(cache.flush, functionCall)
     }
 
-    override def cleanup(flowProcess: FlowProcess[_], operationCall: OperationCall[MapsideCache[V]]) {
+    override def cleanup(flowProcess: FlowProcess[_], operationCall: OperationCall[MapsideCache[Tuple, V]]) {
       // The cache may be large, but super sure we drop any reference to it ASAP
       // probably overly defensive, but it's super cheap.
       operationCall.setContext(null)
     }
   }
 
-  sealed trait MapsideCache[V] {
-    def flush: Option[Map[Tuple, V]]
-    def put(key: Tuple, value: V): Option[Map[Tuple, V]]
+  class TypedMapsideReduce[K, V](
+    @transient fn: TupleEntry => TraversableOnce[(K, V)],
+    @transient commutativeSemigroup: Semigroup[V],
+    sourceFields: Fields,
+    keyFields: Fields, valueFields: Fields,
+    cacheSize: Option[Int])(implicit setKV: TupleSetter[(K, V)])
+    extends BaseOperation[MapsideCache[K, V]](Fields.join(keyFields, valueFields))
+    with Function[MapsideCache[K, V]]
+    with ScaldingPrepare[MapsideCache[K, V]] {
+
+    val boxedSemigroup = Externalizer(commutativeSemigroup)
+    val lockedFn = Externalizer(fn)
+
+    override def prepare(flowProcess: FlowProcess[_], operationCall: OperationCall[MapsideCache[K, V]]) {
+      //Set up the context:
+      implicit val sg: Semigroup[V] = boxedSemigroup.get
+      val cache = MapsideCache[K, V](cacheSize, flowProcess)
+      operationCall.setContext(cache)
+    }
+
+    @inline
+    private def add(evicted: Option[Map[K, V]], functionCall: FunctionCall[MapsideCache[K, V]]) {
+      // Use iterator and while for optimal performance (avoid closures/fn calls)
+      if (evicted.isDefined) {
+        val it = evicted.get.iterator
+        val tecol = functionCall.getOutputCollector
+        while (it.hasNext) {
+          val (key, value) = it.next
+          // Safe to mutate this key as it is evicted from the map
+          tecol.add(setKV(key, value))
+        }
+      }
+    }
+
+    import scala.collection.mutable.{ Map => MMap }
+
+    private[this] class CollectionBackedMap[K, V](val backingMap: MMap[K, V]) extends Map[K, V] with java.io.Serializable {
+      def get(key: K) = backingMap.get(key)
+
+      def iterator = backingMap.iterator
+
+      def +[B1 >: V](kv: (K, B1)) = backingMap.toMap + kv
+
+      def -(key: K) = backingMap.toMap - key
+    }
+
+    private[this] def mergeTraversableOnce[K, V: Semigroup](items: TraversableOnce[(K, V)]): Map[K, V] = {
+      val mutable = MMap[K, V]()
```

Comment:
I've heard java mutable maps are a little faster, might be worth trying that (can use `.asScala` to return the `CollectionBackedMap`. Just an idea.